### ISSUE TITLE
Farm Solution Deployment doesn't work for web app scoped WSP's

### ISSUE THREAD

**Issue by rolandoldengarm** [SubPointSolutions/spmeta2#964]
I've got a couple of WSP's, one of them is Web Application scoped. I've set my model as follows.
````
public static FarmSolutionDefinition UIFarmSolutionDefinition => new FarmSolutionDefinition
        {
            FileName = "Dms.UI.wsp",
            SolutionId = new Guid("xx"),
            ShouldAdd = true,
            ShouldDeploy = true,
            ShouldRetract = true,
            ShouldDelete = true,
            ShouldUpgrade = false
        };
````
I get the following error:
````
There was an error while provisioning definition. Check ModelNode prop. --> This solution contains resources scoped for a  Web application and must be retracted from one or more Web applications.
````
Using SSOM, SharePoint 2016.

I believe FarmSolutionDefinition should have a list of WebApplicationURLs

I remember having a discussion about this on Yammer? 
Is there a workaround? This is blocking us from switching from Powershell SSOM scripts to SPMeta2, hopefully there's a solution. Thanks.

**Comment by SubPointSupport** [comment]
Web app level scope is not supported yet. Only farm level is there. Don't seem to have issues with adding web app level support. That would be extensions to web app syntax (Web App Models). No ETA yet, will update ticket in a few days with more relevant estimation.

**Comment by rolandoldengarm** [comment]
Awesome thanks @SubPointSupport . Looking forward to your ETA, because if this will take too long, I'll have to find an interim solution (i.e. deploy this WSP with PowerShell); but I want to avoid that... You can do it ;)

**Comment by SubPointSupport** [comment]
Have a look on "SPSD SharePoint Solution Deployer". Might be a quick win until we fix web app scope deployment,

**Comment by rolandoldengarm** [comment]
@SubPointSupport any solution for this? Right now, we still use our existing Powershell scripts. Would be cool to kill those.
I don't want to introduce a 3rd party solution.

**Comment by SubPointSupport** [comment]
Next week or early.

**Comment by SubPointSupport** [comment]
In progress.

**Comment by rolandoldengarm** [comment]
@SubPointSupport still in progres...? :)

**Comment by SubPointSupport** [comment]
Yea, sorry. We were really maxed out due to the following issues:
* O365 API regression - Nested terms provision fails #995
* O365 API regression - TermSet not found #994
* O365 API regression - Taxonomy Group provision fails #959

This one is on top now.

**Comment by SubPointSupport** [comment]
Meanwhile, @rolandoldengarm,  is there anything else you'd like to see in the upcoming release?

**Comment by rolandoldengarm** [comment]
Thanks @SubPointSupport 
Probably the only other thing for now is: https://github.com/SubPointSolutions/spmeta2/issues/957

**Comment by SubPointSupport** [comment]
Still, have hesitations - either we should add URLs under FarmSolutionDefinition or enable FarmSolutionDefinition provision under WebApplication.

The question is where should URLs come from? WebApp model or hardcoded under FarmSolutionDefinition?

**Comment by SubPointSupport** [comment]
We'll go with syntax under web application. That was the model wouldn't have hardcoded URLs, it would be possible to reuse and orchestrate correct deployment. URLs would come from the end user while creating web application model.

```cs
var model  = SPMeta2Model.NewWebApplicationModel(webApp => {
    
    webApp.AddFarmSolution(solution);

});

// deploy as usual
```

**Comment by rolandoldengarm** [comment]
yep I like that 👍

**Comment by rolandoldengarm** [comment]
I've tried it, it does not seem to work when the WSP has already been deployed, it does not retract properly. 

Definition:
```
public static FarmSolutionDefinition UIFarmSolutionDefinition => new FarmSolutionDefinition
        {
            FileName = "Dms.UI.wsp",
            SolutionId = new Guid("xyz"),
            ShouldAdd = true,
            ShouldDeploy = true,
            ShouldRetract = true,
            ShouldDelete = true,
            ShouldUpgrade = false,
        };
```
Deployment:
```
var model = SPMeta2Model.NewWebApplicationModel(webApp =>
{
                webApp.AddFarmSolutions(FarmSolutions.UIFarmSolutionDefinition);

            });
```
Error:
```
Exception: There was an error while provisioning definition. Check ModelNode prop. --> This solution contains resources scoped for a Web application and must be retracted from one or more Web applications.
```

If I manually retract it, and run it again, I get this error:
```
Error: 27/04/2017 9:30:38 AM:
Exception: There was an error while provisioning definition. Check
 ModelNode prop. --> An object in the SharePoint administrative framework, "SPSolutionLanguagePack Name=0", depends on other objects which do not exist.  Ensure that all of the objects dependencies are created and retry this operation. -->
The INSERT statement conflicted with the FOREIGN KEY SAME TABLE constraint "FK_Objects_Objects". The conflict occurred i
The statement has been terminated. --> (Line [47] Procedure [proc_putObject] Class [16]  Number [547] State [0] )
```

*I am using incremental provisioning support*.

@SubPointSupport

**Comment by SubPointSupport** [comment]
Hm, we got dozens of web app scope tests for *.wsp provision passed. We'll get back on this one, seems to be a known issue as per the blog posts in the internet.

**Comment by rolandoldengarm** [comment]
So just to be sure, there are 2 issues here. @SubPointSupport

**Comment by SubPointSupport** [comment]
Yes, seems provision waits incorrectly for both deployment and retraction causing havoc. 

Highly suggest to revert to SPSD or PowerShell until this one is solved. Regression on this one is extremely heavy and time consuming. There are 20 plus tests to validate successful deployment of *.wsp package from various stages (deployed, retracted, deleted). Takes quite a time and care.
